### ISSUE TITLE
Configuration results in what I describe as a fork bomb

### ISSUE THREAD

**Issue by gspiegelberg-pivotal** [pgbackrest/pgbackrest#1170]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version:
2.23

2. PostgreSQL version:
11.3

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:
RHEL 7.8 for all hosts (hosts are s8[1-5]), database and repo is same host, standbys exist (5-node patroni cluster) but purposefully removed from config, postgres to postgres ssh password-less capable to all hosts

4. Did you install pgBackRest from source or from a package?
Yes, vmware-postgres-11.3-7 RPM

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
```
[main]
pg1-path=/var/lib/pgsql/11/data
pg1-port=6432
pg1-host=10.2.2.81
pg1-socket-path=/tmp

[global]
repo1-path=/pgbackup
repo1-retention-full=2
start-fast=y

process-max=2

log-level-console=info
log-level-file=debug
log-level-stderr=warn
log-path=/var/lib/pgsql/backup/log
log-subprocess=y
log-timestamp=y
```

    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
From "patronictl show-config pg_uat_cluster_1":
```
loop_wait: 2
maximum_lag_on_failover: 1048576
postgresql:
  parameters:
    archive_mode: 'on'
    checkpoint_timeout: 60
    effective_cache_size: 256MB
    hot_standby: 'on'
    listen_addresses: '*'
    log_connections: true
    log_destination: stderr
    log_directory: log
    log_disconnections: true
    log_filename: postgresql-%a.log
    log_line_prefix: '%m [%p] '
    log_rotation_age: 1d
    log_rotation_size: 0
    log_statement: all
    log_truncate_on_rotation: true
    logging_collector: true
    max_connections: 250
    max_replication_slots: 8
    max_wal_senders: 16
    shared_buffers: 256MB
    synchronous_commit: local
    synchronous_standby_names: ANY 1 (s81,s82,s83)
    wal_keep_segments: 20
    wal_level: replica
  pg_hba:
  - host all         postgres   127.0.0.1/32  trust
  - host all         all        127.0.0.1/32  md5
  - host all         all        10.2.2.0/24   md5
  - host replication replicator 10.2.2.81/32  md5
  - host replication replicator 10.2.2.82/32  md5
  - host replication replicator 10.2.2.83/32  md5
  - host replication replicator 10.2.2.84/32  md5
  - host replication replicator 10.2.2.85/32  md5
  - host replication replicator 10.2.2.87/32  md5
  use_pg_rewind: true
  use_slots: true
retry_timeout: 1
ttl: 9
```
archive_command purposefully excluded from config until I get a successful "check" (see below).

    - errors in the postgresql log file before or during the time you experienced the issue
Only "ERROR" in database log during time of test is:
```
2020-09-10 06:55:51.053 MDT [8439] ERROR:  replication slot "s82" does not exist
```
Believe from when cluster was initialized.  Cluster is healthy.

    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)
Below is repeated countless times:
```
-------------------PROCESS START-------------------
2020-09-10 06:56:52.850 P00   INFO: check:remote command begin 2.23: --log-level-console=off --log-level-file=debug --log-level-s
tderr=error --log-path=/var/lib/pgsql/backup/log --log-subprocess --log-timestamp --pg1-host=10.2.2.81 --pg1-path=/var/lib/pgsql/
11/data --pg1-port=6432 --pg1-socket-path=/tmp --process=0 --remote-type=pg --repo1-path=/pgbackup --stanza=main
2020-09-10 06:56:52.851 P00  DEBUG:     config/load::cfgLoad: => void
2020-09-10 06:56:52.851 P00  DEBUG:     command/remote/remote::cmdRemote: (void)
2020-09-10 06:56:52.851 P00  DEBUG:     protocol/server::protocolServerProcess: (this: {name: remote-0})
2020-09-10 06:56:52.851 P00  DEBUG:     storage/remote/protocol::storageRemoteProtocol: (command: {"dbOpen"}, paramList: null, se
rver: {name: remote-0})
2020-09-10 06:56:52.851 P00  DEBUG:     protocol/helper::pgIsLocal: (hostId: 1)
2020-09-10 06:56:52.851 P00  DEBUG:     protocol/helper::pgIsLocal: => false
2020-09-10 06:56:52.851 P00  DEBUG:     protocol/helper::protocolRemoteGet: (protocolStorageType: 1, hostId: 1)
2020-09-10 06:56:52.851 P00  DEBUG:     protocol/helper::protocolRemoteParam: (protocolStorageType: 1, protocolId: 0, hostIdx: 0)
2020-09-10 06:56:52.852 P00  DEBUG:     protocol/helper::protocolRemoteParam: => {["-o", "LogLevel=error", "-o", "Compression=no"
, "-o", "PasswordAuthentication=no", "postgres@10.2.2.81", "/usr/bin/pgbackrest --log-level-console=off --log-level-file=debug --
log-level-stderr=error --log-subprocess --log-timestamp --pg1-path=/var/lib/pgsql/11/data --pg1-port=6432 --pg1-socket-path=/tmp
--process=0 --remote-type=pg --stanza=main check:remote"]}
2020-09-10 06:56:52.852 P00  DEBUG:     common/exec::execNew: (command: {"ssh"}, param: {["-o", "LogLevel=error", "-o", "Compress
ion=no", "-o", "PasswordAuthentication=no", "postgres@10.2.2.81", "/usr/bin/pgbackrest --log-level-console=off --log-level-file=d
ebug --log-level-stderr=error --log-subprocess --log-timestamp --pg1-path=/var/lib/pgsql/11/data --pg1-port=6432 --pg1-socket-pat
h=/tmp --process=0 --remote-type=pg --stanza=main check:remote"]}, name: {"remote-0 process on '10.2.2.81'"}, timeout: 1830000)
2020-09-10 06:56:52.852 P00  DEBUG:     common/exec::execNew: => {Exec}
2020-09-10 06:56:52.852 P00  DEBUG:     common/exec::execOpen: (this: {Exec})
2020-09-10 06:56:52.853 P00  DEBUG:     common/exec::execOpen: => void
```

7. Describe the issue:
Execution of "check" as postgres user (all path permissions have been checked) on current primary/leader/master host results in pgbackrest in an infinite loop analogous to a fork() bomb until host dies or I kill all pgbackrest processes.  Command used:
```
/usr/bin/pgbackrest --stanza=main --log-level-console=info check
```

With the added postgresql.conf parameter archive_command of:
```
  parameters:
    archive_command: "/usr/bin/pgbackrest --stanza=main archive-push %p"
```
results in database log messages:
```
2020-09-10 08:18:52.060 P00   INFO: archive-push command begin 2.23: [pg_wal/0000002A.history] --log-level-console=info --log-level-file=debug --log-level-stderr=warn --log-path=/var/lib/pgsql/backup/log --log-subprocess --log-timestamp --pg1-host=10.2.2.81 --pg1-path=/var/lib/pgsql/11/data --process-max=2 --repo1-path=/pgbackup --stanza=main
ERROR: [125]: remote-0 process on '10.2.2.81' terminated unexpectedly [255]: Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).
2020-09-10 08:18:52.156 P00   INFO: archive-push command end: aborted with exception [125]
2020-09-10 08:18:52.158 MDT [8433] LOG:  archive command failed with exit code 125
2020-09-10 08:18:52.158 MDT [8433] DETAIL:  The failed archive command was: /usr/bin/pgbackrest --stanza=main archive-push pg_wal/0000002A.history
```

Output of manual execution of archive_command:
```
$ /usr/bin/pgbackrest --stanza=main archive-push pg_wal/0000002A.history
2020-09-10 08:22:05.921 P00   INFO: archive-push command begin 2.23: [pg_wal/0000002A.history] --log-level-console=info --log-level-file=debug --log-level-stderr=warn --log-path=/var/lib/pgsql/backup/log --log-subprocess --log-timestamp --pg1-host=10.2.2.81 --pg1-path=/var/lib/pgsql/11/data --process-max=2 --repo1-path=/pgbackup --stanza=main
ASSERT: [025]: working path '/var/lib/pgsql' is not the same path as '/var/lib/pgsql/11/data'
        STACK TRACE:
        command/archive/common:walPath:278:(debug log level required for parameters)
        command/archive/push/push:cmdArchivePush:(debug log level required for parameters)
        main:main:(debug log level required for parameters)
2020-09-10 08:22:05.921 P00   INFO: archive-push command end: aborted with exception [025]
```

Hopefully just a misconfiguration.

**Comment by dwsteele** [comment]
> Execution of "check" as postgres user ... results in pgbackrest in an infinite loop

This looks like the issue:
```
pg1-host=10.2.2.81
```
You are telling pgBackRest that the pg host is remote when it is actually local. This leads us into an infinite loop in 2.23 which should be fixed in 2.27 by 20d8c76b6. It would still be bad configuration though because you'd be starting a remote worker for no reason.

> ASSERT: [025]: working path '/var/lib/pgsql' is not the same path as '/var/lib/pgsql/11/data'

This means that PostgreSQL and pgBackrest have a different idea about where the PostgreSQL data directory is located. You should make sure that data_directory in PG and pg1-path in pgBackRest are the same.

I do notice, though, that this is being thrown as an assert but it should just be an error. We'll fix that.

**Comment by gspiegelberg-pivotal** [comment]
Confirmed on pg1-host misconfiguration.

ASSERT comparing paths seems to have been brought out by the misconfiguration as well.

**Comment by dwsteele** [comment]
Glad to hear it. I also improved the working dir error message here: #1171.

**Comment by gspiegelberg-pivotal** [comment]
BTW, misconfiguration where backups are meant to be from system A to system A shouldn't result in an infinite loop of SSH madness.  Just saying :). Maybe add a check?  "ERROR: Hey, pg1-host is this host, don't do that"

**Comment by dwsteele** [comment]
As noted above, the infinite loop has been fixed in 2.27.

Figuring out (quickly) that the host is misconfigured can be time connsuming. Especially for archive push/get we don't want to spend the time.

**Comment by dwsteele** [comment]
Though, having said that it might be a good thing to add to the check command where performance is not an issue.
